index.ts

```diff
@@ -104,541 +119,285 @@ class Rabbitr extends EventEmitter {
-                }).catch(error => {
-                  // we were unable to setup a backoff queue or send the message, nack immediately
-                  this.log(`failed to backoff message ${cyan(topic)}`, data, error);
-                  channel.nack(msg);
-                }).asCallback(cb);
-              }
-            }
-          );
-
-          ++ this.pendingMessagesCount;
-
-          // Add a timeout
-          return Bluebird.race([
-            messageAcknowledgement,
-            Bluebird
-              .delay(this.opts.ackWarningTimeout)
-              .then(() => {
-                throw new TimeoutError({isRpc: false, topic});
-              }),
-          ]).catch(TimeoutError, error => {
-            this.emit('warning', {
-              type: 'ack.timeout',
-              queue: topic,
-              message: data,
-            });
-
-            if (this.opts.autoAckOnTimeout === 'acknowledge') {
-              channel.ack(msg);
-            } else if (this.opts.autoAckOnTimeout === 'reject') {
-              channel.nack(msg);
-            }
-            return null;
-          }).then(() => {
-            -- this.pendingMessagesCount;
-            this.postMessage();
-          });
-        };
+      let data = msg.content.toString();
+      if (msg.properties.contentType === 'application/json') {
+        data = parse(data);
+      }
 
-        return Bluebird.fromCallback(callback =>
-          channel.consume(this._formatName(topic), processMessage, {}, callback)
-        ).catch(error => {
-          // istanbul ignore next
-          this.emit('error', error);
-          throw error;
+      this.log(`got a new message on ${cyan(topic)}`, data);
+
+      try {
+        ++ this.pendingMessagesCount;
+
+        // we create a new promise, and make the resolution an 'ack' and the rejection an 'nack'
+        await new Promise(async (ack: () => void, reject) => {
+          const message: Rabbitr.IMessage<TMessage> = {
+            send: this.send.bind(this),
+            rpcExec: this.rpcExec.bind(this),
+            topic,
+            data,
+            channel,
+            ack,
+            reject,
+            properties: msg.properties,
+          };
+
+          executorFunc(message);
         });
-      }).asCallback(cb);
-    });
-  }
 
-  bindExchangeToQueue(exchange: string, queue: string, cb?: Rabbitr.ErrorCallback): Bluebird<void> {
-    // istanbul ignore next
-    if (!this.connectionPromise.isFulfilled()) {
-      // delay until ready
-      return this.whenReady().then(() =>
-        this.bindExchangeToQueue(exchange, queue, cb)
-      );
-    }
+        // if we hit here, we should ack
+        channel.ack(msg);
+      }
+      catch(err) {
+        // if we hit here, we should nack
+        if (!opts.skipBackoff) {
+          // super simple backoff achieved by just delaying performing a #nack
+          await fromCallback(cb => setTimeout(cb, BACKOFF_EXPIRY));
```

Comment:
Common idiom. It could be replaced by wrapper function for example `function wait(miliseconds: number): Promise<void>`.